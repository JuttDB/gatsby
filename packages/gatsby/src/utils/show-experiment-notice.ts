import { getConfigStore } from "gatsby-core-utils"
import reporter from "gatsby-cli/lib/reporter"
import { emitter } from "../redux"
import chalk from "chalk"
import telemetry from "gatsby-telemetry"

type CancelExperimentNoticeCallback = () => void

export type CancelExperimentNoticeCallbackOrUndefined =
  | CancelExperimentNoticeCallback
  | undefined

const ONE_DAY = 24 * 60 * 60 * 1000

interface INoticeObject {
  reason: string
  solution: string
}

type NoticesToShow = INoticeObject & {
  experimentIdentifier: string
}

const noticesToShow: Array<NoticesToShow> = []

export function showExperimentNoticeAfterTimeout(
  experimentIdentifier: string,
  noticeObject: INoticeObject,
  showNoticeAfterMs: number,
  minimumIntervalBetweenNoticesMs: number = ONE_DAY
): CancelExperimentNoticeCallbackOrUndefined {
  const configStoreKey = `lastExperimentNotice.${experimentIdentifier}`

  const lastTimeWeShowedNotice = getConfigStore().get(configStoreKey)

  if (lastTimeWeShowedNotice) {
    if (Date.now() - lastTimeWeShowedNotice < minimumIntervalBetweenNoticesMs) {
      return undefined
    }
  }

  const noticeTimeout = setTimeout(() => {
    noticesToShow.push({ ...noticeObject, experimentIdentifier })

    getConfigStore().set(configStoreKey, Date.now())
  }, showNoticeAfterMs)

  return function clearNoticeTimeout(): void {
    clearTimeout(noticeTimeout)
  }
}

emitter.on(`COMPILATION_DONE`, () => {
  emitter.off(`COMPILATION_DONE`, () => {})

  if (noticesToShow.length > 0) {
    telemetry.trackFeatureIsUsed(`InviteToTryExperiment`)
    let message = `\n\nHello! Your friendly Gatsby maintainers detected ways to improve your site. We're working on new improvements and invite you to try them out *today* and help ready them for general release.`

    noticesToShow.forEach(
      notice =>
        (message += `\n\n${chalk.bgBlue.bold(
          notice.experimentIdentifier
        )}\n-${notice.reason.trim()}\n-${notice.solution.trim()}\n`)
    )

    reporter.info(message)
  }
})
