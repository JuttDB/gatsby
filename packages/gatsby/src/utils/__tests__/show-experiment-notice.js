import { createNoticeMessage } from "../show-experiment-notice"

describe(`show-experiment-notice`, () => {
  it(`generates a message`, () => {
    expect(
      createNoticeMessage([
        {
          noticeText: `hi`,
          umbrellaLink: `http://example.com`,
          experimentIdentifier: `The Flag`,
        },
      ])
    ).toMatchSnapshot()
  })
})
