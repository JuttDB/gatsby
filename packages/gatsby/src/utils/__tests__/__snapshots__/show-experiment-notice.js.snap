// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show-experiment-notice generates a message 1`] = `
"
Hi from the Gatsby maintainers! Based on what we see in your site, these coming
features may help you. All of these can be enabled within gatsby-config.js via
flags (samples below)

[44m[1mThe Flag (http://example.com)[22m[49m, hi
"
`;
